### ISSUE TITLE
Prevent mail sent from Backdrop from being flagged as spam

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1696]
The contact module currently uses the user's email address as 'From' which causes these messages to be marked as spam.

Related issue on drupal.org:  https://www.drupal.org/node/111702

---
PR from @jenlampton at https://github.com/backdrop/backdrop/pull/1852

**Comment by jenlampton** [comment]
Here's a PR that roughly mirrors my patch for Drupal 7. I'm not sure we need to add the setting in backdrop.  To me providing a checkbox that says "Do you want to fix this bug?" is a bit of a cop-out (and an insult to our users). I would love to see that removed, perhaps in 2.x?

**Comment by klonos** [comment]
> To me providing a checkbox that says "Do you want to fix this bug?" is a bit of a cop-out

:+1: for fixing and no option.

**Comment by quicksketch** [comment]
There are a couple of issues in the PR https://github.com/backdrop/backdrop/pull/1282/files

- Code has a `variable_get()` in it yet.
- The IF statement is incredibly difficult to read. Could we reduce the inline assignments and also reduce use of the ternary (`foo ? bar : baz`) operator?
- Needs an update hook.
- Needs tests.

> To me providing a checkbox that says "Do you want to fix this bug?" is a bit of a cop-out

One problem with this is that sometimes checking the box won't fix the bug in the first place. 90% of the time I enter my *personal* email address into the Backdrop installer. I think most people do the same. *This is wrong*. My Backdrop website isn't going to be able to send email as "gmail.com". So even if we had this checkbox, it wouldn't fix the problem if the email address wasn't for the given server.

Perhaps we should handle that separately, but overall I think we should:

- Remove "Site mail" from the installer.
- Default the installer to use `noreply@[current-domain-name.com]` based on the current domain name. Though that might result in domains like `example.dev` if you were doing this locally.

For this issue, couldn't we automatically determine if the @domain.com was a valid sending address somehow? Between the `$_SERVER['HOST_NAME']` and `$_SERVER['HTTP_HOST']` seems like we could make a pretty reasonable guess as to whether the email would be delivered correctly. Perhaps add a variable for "email_domains" that listed additional "safe" domains from which email could be sent, in the event the server has the ability to send as a different domain name, or is using SMTP in order to do so.

**Comment by jenlampton** [comment]
> One problem with this is that sometimes checking the box won't fix the bug in the first place. 90% of the time I enter my personal email address into the Backdrop installer. I think most people do the same. This is wrong. 

I think this is a valid, though separate and unrelated, issue. However:

> Remove "Site mail" from the installer.

yes, definitely yes! 👍 

> Default the installer to use noreply@[current-domain-name.com] based on the current domain name.

Also yes. 👍 

> Though that might result in domains like example.dev if you were doing this locally.

Yes, but if you are doing it locally there is a good chance the email won't be working anyway. (And if it does, hey, I still get emails from my dev sites w/o an issue).

> For this issue, couldn't we automatically determine if the @domain.com was a valid sending address somehow? Between the $_SERVER['HOST_NAME'] and $_SERVER['HTTP_HOST'] seems like we could make a pretty reasonable guess as to whether the email would be delivered correctly.

Okay, but what do we do if it's not valid? Throw something on the status report?

> Perhaps add a variable for "email_domains" that listed additional "safe" domains from which email could be sent, in the event the server has the ability to send as a different domain name, or is using SMTP in order to do so.

Perhaps we can leave this to contrib? SMTP module maybe?

**Comment by jenlampton** [comment]
Filed a new PR at https://github.com/backdrop/backdrop/pull/1823 that doesn't have the setting, removing the need for an upgrade path. https://github.com/backdrop/backdrop/pull/1823

**Comment by Dinornis** [comment]
The current PR test site is initiating install, needs reloading?

**Comment by jenlampton** [comment]
Thanks @Dinornis I've pushed a new PR at https://github.com/backdrop/backdrop/pull/1852

**Comment by quicksketch** [comment]
Reading through the code, this is apparently how it works; with a site mail of `noreply@site.com`:

  - If the domain on `$reply` is the same as the domain on `site_mail`, the `$reply` value is used as the FROM address. No `Reply-To` header is set.
  - If the domain on `$reply_to` is different than the domain of `site_mail`, the string is decompiled and then reassembled, e.g. `First Last <foo@example.com>` becomes `First Last via Site Name <noreply@site.com>`. This is used as the FROM address. The `Reply-To` is set directly from `$reply`.

This seems a bit magical to me.

I also found that this original bug report no longer applies to Backdrop post 1.6. I enabled Contact module and sent an email from my site at `backdrop.dev` (site mail `noreply@backdrop.dev`). Here are the returned headers:

```
Delivered-To: nate@example.com
Received: by 10.28.234.151 with SMTP id g23csp1168691wmi;
        Sun, 30 Apr 2017 16:08:04 -0700 (PDT)
X-Received: by 10.99.62.67 with SMTP id l64mr7709052pga.172.1493593684481;
        Sun, 30 Apr 2017 16:08:04 -0700 (PDT)
Return-Path: <noreply@backdrop.dev>
Received: from natexps.localhost (192.169.22.242.static.etheric.net. [192.169.22.242])
        by mx.google.com with ESMTP id t1si13547717pge.34.2017.04.30.16.08.04
        for <nate@example.com>;
        Sun, 30 Apr 2017 16:08:04 -0700 (PDT)
Received-SPF: neutral (google.com: 192.169.22.242 is neither permitted nor denied by best guess record for domain of noreply@backdrop.dev) client-ip=192.169.22.242;
Authentication-Results: mx.google.com;
       spf=neutral (google.com: 192.169.22.242 is neither permitted nor denied by best guess record for domain of noreply@backdrop.dev) smtp.mailfrom=noreply@backdrop.dev
Received: by natexps.localhost (Postfix, from userid 1000) id 76CDC21614C5; Sun, 30 Apr 2017 16:08:01 -0700 (PDT)
To: nate@example.com
Subject: [Website feedback] Email from site
X-PHP-Originating-Script: 1000:system.mail.inc
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8; format=flowed; delsp=yes
Content-Transfer-Encoding: 8Bit
X-Mailer: Backdrop CMS
Sender: noreply@backdrop.dev
From: noreply@backdrop.dev
Reply-to: bob@example.com
Message-Id: <20170430230801.76CDC21614C5@natexps.localhost>
Date: Sun, 30 Apr 2017 16:08:01 -0700 (PDT)
```

Everything looks pretty much as you would expect. The FROM address is from the site mail, while the Reply-To is the value I entered in the contact form (`bob@example.com`). Note that this still ended up in my Spam folder, because my localhost obviously isn't registered as `backdrop.dev` on the public internet.

So the actual bug in the original report is already fixed. This seems like it's now a request to make it so that if a `$reply` is specified that matches the site mail is passed in, we send directly as that user, a move that may make it more likely to be flagged as spam rather than less.

**Comment by jenlampton** [comment]
Hm. Do you think contact module works differently than the rest of Backdrop? I don't remember any other issues that would have fixed this for core in 1.6 (other than removing site_mail from the installer)

**Comment by quicksketch** [comment]
> Do you think contact module works differently than the rest of Backdrop post 1.6?

Heh, actually looks like we fixed this all the way back in 1.1.0: #305

**Comment by jenlampton** [comment]
Bah. :/
